internal/expr/prepare.go

```diff
@@ -253,47 +366,65 @@ func (pe *ParsedExpr) Prepare(args ...any) (expr *PreparedExpr, err error) {
 		}
 	}
 
-	var sql bytes.Buffer
-
-	var inCount int
-	var outCount int
-
 	var outputs = make([]typeMember, 0)
 	var inputs = make([]typeMember, 0)
 
+	var sqlChunks []sqlChunk
+	var typeMembers []typeMember
+	var chunk sqlChunk
+
 	// Check and expand each query part.
 	for _, part := range pe.queryParts {
 		switch p := part.(type) {
 		case *inputPart:
-			inLoc, err := prepareInput(ti, p)
+			chunk, typeMembers, err = prepareInput(ti, p)
 			if err != nil {
 				return nil, err
 			}
-			sql.WriteString("@sqlair_" + strconv.Itoa(inCount))
-			inCount++
-			inputs = append(inputs, inLoc)
+			inputs = append(inputs, typeMembers...)
 		case *outputPart:
-			outCols, fields, err := prepareOutput(ti, p)
+			chunk, typeMembers, err = prepareOutput(ti, p)
 			if err != nil {
 				return nil, err
 			}
-			for i, c := range outCols {
-				sql.WriteString(c.String())
-				sql.WriteString(" AS ")
-				sql.WriteString(markerName(outCount))
-				if i != len(outCols)-1 {
-					sql.WriteString(", ")
-				}
-				outCount++
+			outputs = append(outputs, typeMembers...)
+		case *inPart:
+			chunk, typeMembers, err = prepareIn(ti, p)
+			if err != nil {
+				return nil, err
 			}
-			outputs = append(outputs, fields...)
-
+			inputs = append(inputs, typeMembers...)
 		case *bypassPart:
-			sql.WriteString(p.chunk)
+			chunk = &bypassChunk{p.chunk}
 		default:
 			return nil, fmt.Errorf("internal error: unknown query part type %T", part)
 		}
+		sqlChunks = append(sqlChunks, chunk)
 	}
 
-	return &PreparedExpr{inputs: inputs, outputs: outputs, sql: sql.String()}, nil
+	return &PreparedExpr{inputs: inputs, outputs: outputs, sqlChunks: sqlChunks}, nil
+}
+
+func (pe *PreparedExpr) sql() string {
+	var inCount int
+	var outCount int
+	var s string
+	var sql bytes.Buffer
+
+	for _, chunk := range pe.sqlChunks {
```

Comment:
As all these implement `sqlChunk`, it feels like this loop could be made more concise:
```go
for _, chunk := range pe.sqlChunks {
    sql.WriteString(chunk.chunk()) // parameters and return values would need to be reworked :)
}
```

Reply:
Changed, a lot shorter now.